### PR TITLE
Parquet metadata

### PR DESCRIPTION
**Pull request by IanCal**
Builds on #2 

The key change here is that there is also an extra file written out.

Parquet files all have a metadata block in them at the end. This allows planners to identify whether or not the file needs to be used, as it contains things like min/max values for columns. It also contains sub-file level detail if there are enough rows.

However, to make use of this a system needs to process all of the files in storage to read the metadata block. Not an issue when all the files are local but a more significant one when they're on S3 as it'd need to crawl everything. In addition we have a non-standard partitioning scheme with multiple levels.

We can collect all of the metadata from each file and put it into one single metadata file that points off to the actual file locations where the data is stored. This should entirely remove the need to generate lists of required files in the other packages (the most common repeated code) and remove the need to pull the files locally before running queries. This is a standard feature in pyarrow, and should be usable outside of our code and outside of just python.

https://arrow.apache.org/docs/python/parquet.html#writing-metadata-and-common-metadata-files

When these are available, we can delete most of https://github.com/cardstack/cardstack/blob/main/packages/cardpay-reward-programs/cardpay_reward_programs/utils.py#L11-L78 and replace it with a single call to pyarrow dataset construction, which we can register in duckdb and remove the parquet_scan manual construction.

### CODE REVIEW

File: subgraph_extractor/cli.py [hunk 1]
```diff
@@ -5,48 +5,52 @@
...
+
+TYPE_MAPPINGS = {"numeric": "bytes", "text": "string", "boolean": "bool", "bytea": "bytes"}
 
 BLOCK_COLUMN = "_block_number"
-BLOCK_COLUMN_TYPE = pyarrow.uint32()
+BLOCK_COLUMN_TYPE = "uint32"
 
 
-def get_select_all_exclusive(
+def select_writable_in_range(
```
Comment (IanCal) [inline subgraph_extractor/cli.py:24]: This is a fix alongside other changes, I'd made a too-complex query to remove some columns when it's far easier just to pull everything and manually remove the columns in the dataframe.

File: subgraph_extractor/cli.py [hunk 2]
```diff
@@ -154,12 +158,20 @@ def convert_columns(df, database_types, table_config):
...
+        # but there is a helper on schemas to create a valid
+        # empty table of the same type
+        table = schema.empty_table()
+    else:
+        table = table.cast(schema, safe=False)
     return table
 
 
-def get_partition_iterator(min_partition, max_partition, partition_sizes):
+def get_partitions(min_partition, max_partition, partition_sizes):
```
Comment (IanCal) [inline subgraph_extractor/cli.py]: Just stopping this being an iterator as all of the uses now require pulling the whole list.

File: subgraph_extractor/cli.py [hunk 3]
```diff
@@ -216,6 +217,61 @@ def write_config(config, root_output_location):
             yaml.dump(config, f_out)
 
 
+def remove_protocol(path):
```
Comment (IanCal) [inline subgraph_extractor/cli.py]: Leaky abstractions around cloud vs local files.

Comment (tintinthong) [reply subgraph_extractor/cli.py]: At first sight, `remove_protocol` don't seem like a good function name. `remove_cloud_prefix` is more direct or anything more approp

**Comment by tintinthong** [comment]
At first sight LGTM, Is there any tests surrounding the newly created metadata file??

**Comment by IanCal** [comment]
> At first sight LGTM, Is there any tests surrounding the newly created metadata file??

Not added, but the tests that were constructing their own partition list & loading that now use the metadata file, so those test that it's written and point to the right data. I'll add in an explicit test that checks that the result of using the metadata file is the same as constructing it manually though, that would help ensure the two approaches remain compatible.

**Comment by IanCal** [comment]
Added an explicit test

File: subgraph_extractor/cli.py (hunk 3 above)
Comment (IanCal) [reply subgraph_extractor/cli.py]: renamed, I like that change

**Comment by IanCal** [comment]
@tintinthong I'm going to cut a release of this if that seems alright. I added an explicit test and the other tests have been modified to use the metadata file for the datasets.

**Review by tintinthong — APPROVED** [review]

**Pull request merged**